components/server/src/jobs/workspace-gc.ts

```diff
@@ -89,18 +89,44 @@ export class WorkspaceGarbageCollector implements Job {
                     this.config.workspaceGarbageCollection.minAgeDays,
                     this.config.workspaceGarbageCollection.chunkLimit,
                 );
+            let deleted = 0;
             const afterSelect = new Date();
             log.info(`workspace-gc: about to soft-delete ${workspaces.length} workspaces`);
             for (const ws of workspaces) {
+                try {
+                    const info = await this.workspaceService.getWorkspace(SYSTEM_USER_ID, ws.id);
+                    let pendingChanges = 0;
+                    if (info.latestInstance) {
+                        pendingChanges += info.latestInstance.gitStatus?.totalUncommitedFiles || 0;
+                        pendingChanges += info.latestInstance.gitStatus?.totalUnpushedCommits || 0;
+                        pendingChanges += info.latestInstance.gitStatus?.totalUntrackedFiles || 0;
+                    }
+                    if (
+                        pendingChanges > 0 &&
+                        info.latestInstance?.creationTime &&
+                        // skip workspaces with pending changes that are younger than minAgeDays times two
+                        new Date(info.latestInstance.creationTime).getTime() >
+                            now.getTime() - this.config.workspaceGarbageCollection.minAgeDays * 2 * 24 * 60 * 60 * 1000
+                    ) {
+                        log.info(
+                            { workspaceId: ws.id, pendingChanges, lastUse: info.latestInstance.creationTime },
+                            "workspace-gc: skipping workspace with pending changes",
+                        );
+                        continue;
```

Comment:
Without ordering by workspaceAge DESC, this can lead to us getting stuck internally on a chunk of workspaces that won't ever be deleted.

Reply:
Alternatively, we could introduce a new state "gc-delayed" and modify [this query](https://github.com/gitpod-io/gitpod/pull/19819/files#diff-9a9f7652c2062d548972139aecbcdd8ccde1ca1218ea6a898866db13f765b852R146) to make it look for it.
Honestly liking that a tad better, because it's more explicit :eyes:

Reply:
Not sure I understand. The user should see the workspace for longer, so we want to postpone the softdeletion not the harddeletion.

Reply:
> The user should see the workspace for longer, so we want to postpone the softdeletion not the harddeletion.

The idea is to introduce another (optional) step in the state machine. But it's true, that would require more work.

Sorting should be easier :wink:

Reply:
You mean updating a persistent field `hasGitChanges` on workspace whenever gitstatus gets updated and then using that flag in the DB query?

Reply:
Anyway, I push ordering by lastUse descending as another commit. 
Is that what you meant?

Reply:
> Is that what you meant?

:dart: Yes, that should work. :+1:

Reply:
> You mean updating a persistent field hasGitChanges on workspace whenever gitstatus gets updated and then using that flag in the DB query?

For posterity: Tha's one solution, yes. I would have tried to go with a SQL expression like `SUM(gitStatus->>'$.uncommitedFilesTotal', gitStatus->>'$.totalUntrackedFiles', gitStatus->>'$.totalUnpushedCommits')` first: That should be reasonably fast given the pre-selection we are already doing.

Next step would be to mark it as `gc-with-changes`.
And have a separate step+query for that in the WorkspaceGC.